README.md
# Terrain navigator

- get height data from mapbox API by coordinates
- cache height data in IndexDB
- create terrains far from observer position
- resize imageData for low-poly tiles
- [optimize rendering at very far distances](https://discourse.threejs.org/t/moving-the-camera-model-will-shake-if-the-coordinates-are-large/7214)
- rebuild tiles after change observer position

## Quick start

```js
import TerrainNavigator from 'terrain-navigator';

const scene: THREE.Scene;
const observer: THREE.Object3D;

const terrain = new TerrainNavigator({
  mapBoxToken: MAPBOX_TOKEN,
  container: scene,
  material: new MeshPhongMaterial({
    color: 0x424240,
    shininess: 0.1,
    // wireframe: true,
  });,
  zoom: 14,
  scale: 0.125,
  getPosition() {
    return observer.target.position;
  },
  setPosition(pos) {
    const { position } = observer.object;

    Object.entries(pos).forEach(([key, val]) => {
      if (Number.isFinite(val)) position[key] = val;
    });
  },
  onTileRebuilded(tile, oldObject) {
    if (oldObject) observer.removeTeleportTargets([oldObject]);

    scene.add(tile.object);

    if (tile.isNear) observer.addTeleportTargets([tile.object]);

    editor.dyeTile(tile);
  },
});

// every time you change the observer position:
terrain.update();
```

TODO:

- move data processing to Worker


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.0.1](https://github.com/apostololeg/terrain-navigator/compare/v2.0.0...v2.0.1) (2023-08-20)


### Features

* Multiple levels support ([a1bcfb1](https://github.com/apostololeg/terrain-navigator/commit/a1bcfb1bb09cd0ad63f264eea8df9c4e0637c130))
* seam to prev level (top, bottom) ([dba0284](https://github.com/apostololeg/terrain-navigator/commit/dba0284a3abadf27a33366684cef9da14169d0d1))
* tile.clipSide ([bbbf782](https://github.com/apostololeg/terrain-navigator/commit/bbbf782c4f3513e23a6df96a51e8869763352b25))


### Bug Fixes

* clippingPlanes ([2320457](https://github.com/apostololeg/terrain-navigator/commit/23204579398c80db60e199fda4e7e80a39ac6e7f))

## [2.0.0](https://github.com/apostololeg/terrain-navigator/compare/v1.0.1...v2.0.0) (2022-11-12)


### Features

* Novemtree => Quadtree ([a35844b](https://github.com/apostololeg/terrain-navigator/commit/a35844bd475c96048629e98aa7999ffe8d5fba7b))

### [1.0.1](https://github.com/apostololeg/terrain-navigator/compare/v1.0.0...v1.0.1) (2022-10-26)


### Bug Fixes

* seamTiles - copyNormals ([7a8ea70](https://github.com/apostololeg/terrain-navigator/commit/7a8ea70c51e696458fade968edc196e1e1e031fe))

## 1.0.0 (2022-10-25)


src/tiles.ts
import Dexie from 'dexie';
import resizeImageData from 'resize-image-data';

import { blobToImageData, imageToHeightData } from './utils';

export type HeightData = number[];

const _ = (...args) => args.join('_');
const db = new Dexie('heightmap');

db.version(2).stores({
  height: 'slug,data',
  image: 'slug,data',
});

const init = db.open();

export default class Tiles {
  params;

  constructor(params) {
    this.params = params;
  }

  async getTile(x, z, zoom, size: number): Promise<HeightData> {
    await init;

    let slug = _(x, z, zoom, size);
    // if (zoom === 12) await db.table('height').delete(slug);
    let item = await db.table('height').get({ slug });
    let data = item?.data;

    if (!data) {
      const imageData = await this.getImageData(x, z, zoom, size);
      data = imageToHeightData(imageData as ImageData);
      db.table('height').put({ slug, data });
    }

    return data;
  }

  async getImageData(x, z, zoom, size) {
    const imageData = await this.loadImageData(x, z, zoom);

    return size !== imageData.height
      ? resizeImageData(imageData, size, size)
      : imageData;
  }

  buildTileURL = (x, y, zoom, tileset) => {
    const token = this.params.mapBoxToken;
    return `https://api.mapbox.com/v4/mapbox.${tileset}/${zoom}/${x}/${y}.pngraw?access_token=${token}`;
  };

  async loadImageData(x, z, zoom): Promise<ImageData> {
    const cache = await db.table('image').get({ slug: _(x, z, zoom) });
    if (cache) return cache.data;

    const url = this.buildTileURL(x, z, zoom, 'terrain-rgb');
    const res = await fetch(url);
    const blob = await res.blob();
    const data = await blobToImageData(blob);
    const slug = _(x, z, zoom);

    db.table('image').put({ slug, data });

    return data;
  }
}


src/utils.ts
import { Vector2 } from 'three';
import { yAmplitude, yOffset, SIDE, Side, TILE_IMAGE_SIZE } from './constants';

export async function blobToImageData(blob) {
  let blobUrl = URL.createObjectURL(blob);
  const size = TILE_IMAGE_SIZE;

  return new Promise((resolve, reject) => {
    const img = new Image(size, size);

    img.onload = () => resolve(img);
    img.onerror = err => reject(err);
    img.src = blobUrl;
  }).then((img: HTMLImageElement) => {
    URL.revokeObjectURL(blobUrl);
    const canvas = document.createElement('canvas');
    canvas.width = size;
    canvas.height = size;
    const ctx = canvas.getContext('2d');
    ctx.drawImage(img, 0, 0);
    return ctx.getImageData(0, 0, size, size); // some browsers synchronously decode image here
  });
}

export function imageToHeightData(imageData: ImageData): number[] {
  const { data, width, height } = imageData;
  const heightData = new Array(width * height);

  for (let i = 0; i < data.length; i += 4) {
    const R = data[i];
    const G = data[i + 1];
    const B = data[i + 2];
    const height = (R * 256 * 256 + G * 256 + B) * yAmplitude - yOffset;

    heightData[i / 4] = +height.toFixed(2);
  }

  return heightData;
}

export function setVerticesData(geometry, data) {
  const vertices = geometry.attributes.position.array;
  const sideSize = Math.sqrt(data.length);
  let lastIndex;
  let i = 0;

  for (let x = 0; x < sideSize; x++) {
    for (let y = 0; y < sideSize; y++) {
      lastIndex = x * sideSize + y;
      const height = data[lastIndex];
      vertices[i * 3 + 2] = height;
      i++;
    }
  }

  // console.log(`sideSize=${sideSize} lastIndex=${lastIndex} \n\n`);
}

export function getClosestCorner(p1, p2, halfTileSizeInMeters): Side {
  const leftX = p1.x - halfTileSizeInMeters;
  const rightX = p1.x + halfTileSizeInMeters;
  const topZ = p1.z - halfTileSizeInMeters;
  const bottomZ = p1.z + halfTileSizeInMeters;
  const point = new Vector2(p2.x, p2.z);
  const distances = {
    [SIDE.TOP_LEFT]: point.distanceToSquared(new Vector2(leftX, topZ)),
    [SIDE.TOP_RIGHT]: point.distanceToSquared(new Vector2(rightX, topZ)),
    [SIDE.BOTTOM_LEFT]: point.distanceToSquared(new Vector2(leftX, bottomZ)),
    [SIDE.BOTTOM_RIGHT]: point.distanceToSquared(new Vector2(rightX, bottomZ)),
  };

  // @ts-ignore
  const [side] = Object.entries(distances).sort((a, b) => a[1] - b[1])[0];

  return side;
}

export function shiftArr(arr: any[], delta, newItem) {
  let oldItem;

  if (delta > 0) {
    oldItem = arr.shift();
    arr.push(newItem);
  } else {
    oldItem = arr.pop();
    arr.unshift(newItem);
  }

  return oldItem;
}


src/constants.ts
export const TILE_IMAGE_SIZE = 256;
export const TILE_IMAGE_SIZE_IN_METERS = TILE_IMAGE_SIZE;
export const yAmplitude = 0.04;
export const yOffset = 0;

export const SIDE = {
  CENTER: 'center',
  TOP: 'top',
  TOP_LEFT: 'top-left',
  TOP_RIGHT: 'top-right',
  RIGHT: 'right',
  LEFT: 'left',
  BOTTOM: 'bottom',
  BOTTOM_RIGHT: 'bottom-right',
  BOTTOM_LEFT: 'bottom-left',
};

export type Side = typeof SIDE[keyof typeof SIDE];

const {
  TOP,
  TOP_LEFT,
  TOP_RIGHT,
  RIGHT,
  LEFT,
  BOTTOM,
  BOTTOM_LEFT,
  BOTTOM_RIGHT,
} = SIDE;

export const TILE_SIDES = [TOP_LEFT, TOP_RIGHT, BOTTOM_LEFT, BOTTOM_RIGHT];

export const TILE_DELTA_BY_SIDE = {
  [TOP_LEFT]: {
    [TOP_LEFT]: [0, 0],
    [TOP_RIGHT]: [0, 1],
    [BOTTOM_LEFT]: [1, 0],
    [BOTTOM_RIGHT]: [1, 1],
  },
  [TOP_RIGHT]: {
    [TOP_LEFT]: [0, -1],
    [TOP_RIGHT]: [0, 0],
    [BOTTOM_LEFT]: [1, -1],
    [BOTTOM_RIGHT]: [1, 0],
  },
  [BOTTOM_LEFT]: {
    [TOP_LEFT]: [-1, 0],
    [TOP_RIGHT]: [-1, 1],
    [BOTTOM_LEFT]: [0, 0],
    [BOTTOM_RIGHT]: [0, 1],
  },
  [BOTTOM_RIGHT]: {
    [TOP_LEFT]: [-1, -1],
    [TOP_RIGHT]: [-1, 0],
    [BOTTOM_LEFT]: [0, -1],
    [BOTTOM_RIGHT]: [0, 0],
  },
};

export const TILE_SEAMS = {
  [TOP_LEFT]: [TOP, LEFT],
  [TOP_RIGHT]: [TOP, RIGHT],
  [BOTTOM_RIGHT]: [BOTTOM, RIGHT],
  [BOTTOM_LEFT]: [BOTTOM, LEFT],
};

export const SEAM_TILES = {
  // [tileSide]: [tileASide, tileBSide]
  [TOP]: [TOP_LEFT, TOP_RIGHT],
  [RIGHT]: [TOP_RIGHT, BOTTOM_RIGHT],
  [BOTTOM]: [BOTTOM_LEFT, BOTTOM_RIGHT],
  [LEFT]: [TOP_LEFT, BOTTOM_LEFT],
};

export const OPPOSITE_SIDE = {
  [TOP_LEFT]: BOTTOM_RIGHT,
  [TOP_RIGHT]: BOTTOM_LEFT,
  [BOTTOM_LEFT]: TOP_RIGHT,
  [BOTTOM_RIGHT]: TOP_LEFT,
};


src/index.ts
import {
  Mesh,
  Group,
  Object3D,
  Matrix4,
  Vector3,
  Raycaster,
  Material,
  PlaneGeometry,
  Plane,
} from 'three';
import SphericalMercator from '@mapbox/sphericalmercator';

import {
  TILE_IMAGE_SIZE,
  TILE_IMAGE_SIZE_IN_METERS,
  TILE_DELTA_BY_SIDE,
  TILE_SIDES,
  TILE_SEAMS,
  SEAM_TILES,
  SIDE,
} from './constants';
import type { Side } from './constants';
import { getClosestCorner, setVerticesData, shiftArr } from './utils';
import Tiles from './tiles';
import type { HeightData } from './tiles';
import { seamSameLevel, seamToPrevLevel } from './seam';

type TilePos = {
  x: number;
  z: number;
  nx: number;
  nz: number;
};

type TileState = {
  side?: Side;
  size?: number;
  scale: number;
  pos?: TilePos;
  heightData?: HeightData;
  geometry: PlaneGeometry;
  material?: Material;
  clipSide?: Side;
  clippingPlanes?: Plane[];
  object?: Object3D<Mesh<PlaneGeometry, Material>>;
  isNear?: boolean;
  // needsRebuild?: boolean;
};

const _t = (): TileState => ({
  geometry: null,
  scale: 0,
  // needsRebuild: true,
});

type Coords = { lat: number; lon: number };

type Vector2D = { x: number; z: number };

export type TerrainParams = {
  mapBoxToken: string;
  container: Object3D;
  material: Material;
  coords?: Coords;
  tileImageSize?: 256 | 512;
  zoom: number;
  minZoom?: number;
  maxZoom?: number;
  zoomScale?: number;
  levelNumber?: number;
  position?: Vector2D;
  offset: Vector2D;
  scale?: number; // 0..1
  getPosition: () => Vector3;
  setPosition: (position: Partial<Vector3>) => void;
  onTileRebuilded?: (
    levelNumber: number,
    tile: TileState,
    oldObject?: Object3D
  ) => void;
};

export default class Terrain {
  _?: Terrain; // root
  parent?: Terrain;
  params?: TerrainParams;
  isRoot: boolean;

  container = new Group();
  merc = new SphericalMercator({
    size: TILE_IMAGE_SIZE,
    // antimeridian: true,
  });

  inited = false;
  scale = 1;
  zoomScale = 1;
  levelNumber = 0;
  subLevel: Terrain;
  offset = { x: 0, z: 0 };
  position = { x: 0, z: 0 }; // observer position with offset
  coords = { lat: 0, lon: 0 };
  parentLevelCenterDot: number[];
  material: Material;
  tileNumber = { x: 0, z: 0 };
  tileSegmentsCount: number;
  imageSize: number;
  tileImageSize: number;
  tileSizeInMeters: number;
  halfTileSizeInMeters: number;
  tilesStore;
  tileBySide = {}; // [side]: TileState
  tilesDeltas = {}; // [side]: [dx, dz]
  tiles = [
    [_t(), _t()],
    [_t(), _t()],
  ];
  currTileSide: string = null;
  currTile: TileState = null;
  rebuildId;
  rebuildPromises: { [side: Side]: Promise<void> } = {}; // [side]: Promise<void>
  rebuildSeams: { [side: Side]: true } = {}; // [seam position]: true
  seamsRebuilded = 0;

  log: (...args: any[]) => void;

  constructor(params: TerrainParams, root?: Terrain, parent?: Terrain) {
    const {
      coords,
      position,
      tileImageSize,
      zoom,
      scale,
      levelNumber,
      mapBoxToken,
    } = params;
    const minZoom = params.minZoom || zoom;

    this.isRoot = !root;
    this._ = root || this;
    this.parent = parent;
    this.params = params;
    if (scale) this.scale = scale;

    if (this.isRoot) {
      if (position) {
        const [lon, lat] = this.merc.ll(position, zoom);
        this.coords = { lat, lon };
      } else if (!coords) {
        throw new Error('coords or position is required');
      } else {
        this.coords = { lat: coords.lat, lon: coords.lon };
      }

      this.tilesStore = new Tiles({ mapBoxToken });
      this.tileImageSize = tileImageSize || TILE_IMAGE_SIZE;
    }

    this.levelNumber = levelNumber ?? 0;
    this.zoomScale = Math.pow(2, this.levelNumber);
    this.imageSize = this._.tileImageSize * this.zoomScale;
    this.tileSegmentsCount =
      (this._.tileImageSize / (this.zoomScale / 2)) * this.scale + 1;

    this.log = console.log.bind(this, `${this.levelNumber} ::`);

    if (typeof minZoom === 'number') {
      const newZoom = zoom - 1;

      this.tileSizeInMeters = TILE_IMAGE_SIZE_IN_METERS * this.zoomScale;
      this.halfTileSizeInMeters = this.tileSizeInMeters / 2;

      if (newZoom >= minZoom) {
        const newLevelParams = {
          ...params,
          zoom: newZoom,
          levelNumber: this.levelNumber + 1,
        };

        this.subLevel = new Terrain(newLevelParams, this._, this);
      }
    }
  }

  async start() {
    if (this.isRoot) this.updateHorizontalPosition();
    await this.update();
    if (this.isRoot) this.updateVerticalPosition();
    this.inited = true;
  }

  updateHorizontalPosition() {
    const { lat, lon } = this._.coords;
    const [x, z] = this.merc.px([lon, lat], this.params.zoom);

    // offset fix objects shaking when camera moves
    // https://discourse.threejs.org/t/moving-the-camera-model-will-shake-if-the-coordinates-are-large/7214
    this.offset = { x, z };
    this._.params.setPosition({ x: 0, z: 0 });
  }

  updateVerticalPosition() {
    const ray = new Raycaster(new Vector3(0, 10000, 0), new Vector3(0, -1, 0));
    const intersectObj = ray.intersectObject(this.currTile.object)[0];
    const yPos = intersectObj?.point.y;

    if (Number.isFinite(yPos)) this.params.setPosition({ y: yPos });
  }

  shift() {
    const { x, z } = this._.position;
    const nx = Math.floor(x / this.imageSize);
    const nz = Math.floor(z / this.imageSize);
    let dx = 0;
    let dz = 0;

    if (this.inited && (this.tileNumber.x !== nx || this.tileNumber.z !== nz)) {
      dx = nx - this.tileNumber.x;
      dz = nz - this.tileNumber.z;
    }

    this.tileNumber = { x: nx, z: nz };

    // counting that dx and dz can be 0 or 1 or -1

    if (dx) {
      this.tiles.forEach(xRow => {
        const oldTile = shiftArr(xRow, dx, _t());
        this.removeTile(oldTile);
      });
    }

    if (dz) {
      const newRow = [_t(), _t()];
      const oldRow = shiftArr(this.tiles, dz, newRow);

      oldRow.forEach(this.removeTile);
    }
  }

  update = async (centerDot?: number[]) => {
    this.log('centerDot', centerDot);

    const { x, z } = this.params.getPosition();

    if (this.isRoot) {
      const { offset } = this._;

      this.position = {
        x: x + offset.x,
        z: z + offset.z,
      };
    }

    if (centerDot) {
      this.parentLevelCenterDot = centerDot;
    }

    this.shift();
    await this.rebuild();

    if (this.subLevel) {
      await this.rebuildPromises[SIDE.TOP_LEFT];
      const leftTopTile = this.tileBySide[SIDE.TOP_LEFT];
      const centralTilePos = [
        leftTopTile.pos.x + this.halfTileSizeInMeters,
        leftTopTile.pos.z + this.halfTileSizeInMeters,
      ];

      this.subLevel.update(centralTilePos);
    }
  };

  async rebuild(): Promise<void[]> {
    const id = Date.now();
    const pos = this.params.getPosition();
    const currTilePos = this.getTilePos(0, 0);
    const newSide = getClosestCorner(
      pos,
      currTilePos,
      this.halfTileSizeInMeters
    );
    const tiles = this.tiles.flat();

    this.log('newSide', newSide);

    if (this.isRoot) this.rebuildId = Date.now();

    if (this.currTileSide === newSide) {
      tiles.forEach(this.updateClipingPlanes);
      if (!this.isRoot) this.seamToPrevLevel();
      return;
    }

    this.currTileSide = newSide;
    this.tilesDeltas = TILE_DELTA_BY_SIDE[newSide];

    const sides = [...TILE_SIDES];
    tiles.forEach(tile => {
      tile.side = sides.shift(); // assign side
      this.tileBySide[tile.side] = tile;
    });
    this.currTile = this.tileBySide[newSide];

    // @ts-ignore
    Object.entries(this.tilesDeltas).forEach(([side, [dx, dz]]) => {
      this.rebuildPromises[side] = this.rebuildTile(dz, dx, side);
    });

    this.seamsRebuilded = 0;
    await Promise.all(Object.values(this.rebuildPromises));
  }

  async rebuildTile(dx, dz, side) {
    const id = this._.rebuildId;
    const { onTileRebuilded } = this.params;
    const tile = this.tileBySide[side];
    const oldObject = tile.object;

    // if (tile.needsRebuild) {
    this.log('rebuildTile', tile.side);

    const pos = this.getTilePos(dx, dz);

    tile.pos = pos;

    this.setTileMaterial(tile);
    this.updateClipingPlanes(tile);

    tile.heightData = await this.getTile(pos.nx, pos.nz);

    if (id !== this._.rebuildId) return;

    tile.geometry = this.buildTileGeometry(pos, tile.heightData);

    this.removeTile(tile);

    tile.object = new Mesh(tile.geometry, tile.material);
    tile.object.receiveShadow = true;
    tile.object.castShadow = true;

    // delete tile.needsRebuild;
    onTileRebuilded(this.levelNumber, tile, oldObject);
    // }

    TILE_SEAMS[side].map(this.seam);
  }

  buildTileGeometry(pos: TilePos, heightData) {
    const segmentCount = Math.sqrt(heightData.length) - 1;
    const geometry = new PlaneGeometry(
      this.tileSizeInMeters,
      this.tileSizeInMeters,
      segmentCount,
      segmentCount
    );

    setVerticesData(geometry, heightData);
    geometry.applyMatrix4(new Matrix4().makeRotationX(-Math.PI / 2));
    geometry.applyMatrix4(new Matrix4().makeTranslation(pos.x, 0, pos.z));
    geometry.computeVertexNormals();

    return geometry;
  }

  removeTile = (tile: TileState) => {
    if (tile.object) {
      this.params.container.remove(tile.object);
      tile.object = null;
    }
  };

  getTile(nx, nz) {
    const { zoom } = this.params;
    const size = this.tileSegmentsCount;

    return this._.tilesStore.getTile(nx, nz, zoom, size);
  }

  getTilePos(dx, dz): TilePos {
    const nx = this.tileNumber.x + dx;
    const nz = this.tileNumber.z + dz;
    const { offset } = this._;

    return {
      nx,
      nz,
      x: nx * this.tileSizeInMeters - offset.x + this.halfTileSizeInMeters,
      z: nz * this.tileSizeInMeters - offset.z + this.halfTileSizeInMeters,
    };
  }

  getClipingPlanes(tile: TileState): Plane[] {
    const { clipSide, pos } = tile;

    if (!clipSide) return [];

    switch (clipSide) {
      case SIDE.CENTER:
        switch (tile.side) {
          case SIDE.TOP_LEFT:
            return [
              new Plane(
                new Vector3(-1, 0, 0),
                -Math.abs(pos.x) - this.halfTileSizeInMeters
              ),
            ];
          case SIDE.TOP_RIGHT:
            return [
              new Plane(
                new Vector3(1, 0, 0),
                -Math.abs(pos.x) - this.halfTileSizeInMeters
              ),
            ];
          case SIDE.BOTTOM_LEFT:
            return [
              new Plane(
                new Vector3(0, 0, -1),
                -Math.abs(pos.z) - this.halfTileSizeInMeters
              ),
            ];
          case SIDE.BOTTOM_RIGHT:
            return [
              new Plane(
                new Vector3(0, 0, 1),
                -Math.abs(pos.z) - this.halfTileSizeInMeters
              ),
            ];
        }

        break;

      case SIDE.RIGHT:
        return [new Plane(new Vector3(-1, 0, 0), pos.x)];

      case SIDE.LEFT:
        return [new Plane(new Vector3(1, 0, 0), -pos.x)];

      case SIDE.BOTTOM:
        return [new Plane(new Vector3(0, 0, -1), pos.z)];

      case SIDE.TOP:
        return [new Plane(new Vector3(0, 0, 1), -pos.z)];

      case SIDE.BOTTOM_RIGHT:
        return [
          new Plane(new Vector3(-1, 0, 0), pos.x),
          new Plane(new Vector3(0, 0, -1), pos.z),
        ];

      case SIDE.BOTTOM_LEFT:
        return [
          new Plane(new Vector3(1, 0, 0), -pos.x),
          new Plane(new Vector3(0, 0, -1), pos.z),
        ];

      case SIDE.TOP_RIGHT:
        return [
          new Plane(new Vector3(-1, 0, 0), pos.x),
          new Plane(new Vector3(0, 0, 1), -pos.z),
        ];

      case SIDE.TOP_LEFT:
        return [
          new Plane(new Vector3(1, 0, 0), -pos.x),
          new Plane(new Vector3(0, 0, 1), -pos.z),
        ];
    }

    return [];
  }

  getClipSide = (tile: TileState) => {
    if (!this.parentLevelCenterDot) return null;

    const { pos } = tile;
    const [x, z] = this.parentLevelCenterDot;
    const dx = pos.x - x;
    const dz = pos.z - z;

    const sameHoriz = dx === 0;
    const sameVert = dz === 0;
    const halfToLeft = dx === -this.halfTileSizeInMeters;
    const halfToRight = dx === this.halfTileSizeInMeters;
    const halfToTop = dz === -this.halfTileSizeInMeters;
    const halfToBottom = dz === this.halfTileSizeInMeters;

    if (sameHoriz && sameVert) return SIDE.CENTER;

    if (sameVert) {
      if (halfToLeft) return SIDE.RIGHT;
      if (halfToRight) return SIDE.LEFT;
    }

    if (sameHoriz) {
      if (halfToTop) return SIDE.BOTTOM;
      if (halfToBottom) return SIDE.TOP;
    }

    if (halfToTop) {
      if (halfToLeft) return SIDE.BOTTOM_RIGHT;
      if (halfToRight) return SIDE.BOTTOM_LEFT;
    }

    if (halfToBottom) {
      if (halfToLeft) return SIDE.TOP_RIGHT;
      if (halfToRight) return SIDE.TOP_LEFT;
    }

    return null;
  };

  setTileMaterial(tile: TileState) {
    if (tile.material) return;

    tile.material = this._.params.material.clone();
    // tile.material.clippingPlanes = tile.clippingPlanes;
    // tile.material.clipIntersection = true;
  }

  updateClipingPlanes = (tile: TileState) => {
    if (this.levelNumber === 0) return;

    tile.clipSide = this.getClipSide(tile);
    tile.clippingPlanes = this.getClipingPlanes(tile);
    this.applyClipingPlanes(tile);
    // console.log('updateClipingPlanes', tile.side, tile.clippingPlanes);
    // if (!tile.clippingPlanes) debugger;
  };

  applyClipingPlanes = (tile: TileState) => {
    if (this.levelNumber === 0) return;

    const { material, clippingPlanes } = tile;

    if (clippingPlanes) {
      material.clippingPlanes = clippingPlanes;
      material.clipIntersection = true;
    } else {
      material.clippingPlanes = [];
      material.clipIntersection = false;
    }
  };

  seam = async seam => {
    if (this.rebuildSeams[seam]) return;

    this.rebuildSeams[seam] = true;

    const id = this._.rebuildId;
    const tilesSides = SEAM_TILES[seam];

    // await for tiles (that adjacent to this seam) to rebuild
    await Promise.all(tilesSides.map(side => this.rebuildPromises[side]));

    if (id !== this._.rebuildId) return;

    const [tileA, tileB] = tilesSides.map(side => this.tileBySide[side]);

    // this.log('seam', seam, '|', tileA.side, tileB.side);
    seamSameLevel(tileA, tileB);

    delete this.rebuildSeams[seam];

    if (!this.isRoot && ++this.seamsRebuilded === 4) {
      this.seamToPrevLevel();
    }
  };

  seamToPrevLevel() {
    let i = 0;
    const id = this._.rebuildId;
    const tiles = this.tiles.flat();

    const flush = () => {
      if (id !== this._.rebuildId) return;
      seamToPrevLevel(tiles[i], this);
      tiles[++i] ? flush() : this.computeVertexNormals();
      // requestAnimationFrame(tiles[++i] ? flush : this.computeVertexNormals);
    };

    flush();
  }

  computeVertexNormals = () => {
    this.tiles
      .flat()
      // @ts-ignore
      .forEach(tile => tile.object.geometry.computeVertexNormals());
  };
}
